File: Provenance/Controller/PVVBControllerViewController.m
//
//  PVVBControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 12/21/16.
//  Copyright © 2016 James Addyman. All rights reserved.
//

#import "PVVBControllerViewController.h"
#import <PVMednafen/MednafenGameCore.h>

@interface PVVBControllerViewController ()

@end

@implementation PVVBControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:OEVBButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:OEVBButtonB];
        }
    }
    
    [self.leftShoulderButton setTag:OEVBButtonL];
    [self.rightShoulderButton setTag:OEVBButtonR];

    [self.selectButton setTag:OEVBButtonSelect];
    [self.startButton setTag:OEVBButtonStart];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    
    [vbCore didReleaseVBButton:OEVBButtonLeftUp forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftDown forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftLeft forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [vbCore didPushVBButton:OEVBButtonLeftUp forPlayer:0];
            [vbCore didPushVBButton:OEVBButtonLeftLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [vbCore didPushVBButton:OEVBButtonLeftUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [vbCore didPushVBButton:OEVBButtonLeftUp forPlayer:0];
            [vbCore didPushVBButton:OEVBButtonLeftRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [vbCore didPushVBButton:OEVBButtonLeftLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [vbCore didPushVBButton:OEVBButtonLeftRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [vbCore didPushVBButton:OEVBButtonLeftDown forPlayer:0];
            [vbCore didPushVBButton:OEVBButtonLeftLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [vbCore didPushVBButton:OEVBButtonLeftDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [vbCore didPushVBButton:OEVBButtonLeftDown forPlayer:0];
            [vbCore didPushVBButton:OEVBButtonLeftRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    
    [vbCore didReleaseVBButton:OEVBButtonLeftUp forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftDown forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftLeft forPlayer:0];
    [vbCore didReleaseVBButton:OEVBButtonLeftRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didPushVBButton:[button tag] forPlayer:0];
    
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didReleaseVBButton:[button tag] forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didPushVBButton:OEVBButtonStart forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didReleaseVBButton:OEVBButtonStart forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didPushVBButton:OEVBButtonSelect forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *vbCore = (MednafenGameCore *)self.emulatorCore;
    [vbCore didReleaseVBButton:OEVBButtonSelect forPlayer:player];
}

@end

File: Provenance/Controller/PVVBControllerViewController.swift
//
//  PVVBControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright © 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVVBButton {
        get {
            return PVVBButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVVBControllerViewController: PVControllerViewController<PVVirtualBoySystemResponderClient> {
    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "A" {
                button.buttonTag = .a
            } else if title == "B" {
                button.buttonTag = .b
            }
        }

        leftShoulderButton?.buttonTag = .l
        rightShoulderButton?.buttonTag = .r
        selectButton?.buttonTag = .select
        startButton?.buttonTag = .start
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.leftUp, forPlayer: 0)
        emulatorCore.didRelease(.leftDown, forPlayer: 0)
        emulatorCore.didRelease(.leftLeft, forPlayer: 0)
        emulatorCore.didRelease(.leftRight, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.leftUp, forPlayer: 0)
                emulatorCore.didPush(.leftLeft, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.leftUp, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.leftUp, forPlayer: 0)
                emulatorCore.didPush(.leftRight, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.leftLeft, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.leftRight, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.leftDown, forPlayer: 0)
                emulatorCore.didPush(.leftLeft, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.leftDown, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.leftDown, forPlayer: 0)
                emulatorCore.didPush(.leftRight, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.leftUp, forPlayer: 0)
        emulatorCore.didRelease(.leftDown, forPlayer: 0)
        emulatorCore.didRelease(.leftLeft, forPlayer: 0)
        emulatorCore.didRelease(.leftRight, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.start, forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.start, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.select, forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.select, forPlayer: player)
    }
}
